File: frontend/src/types/track.ts
import { WritableSignal } from "@angular/core";
import {
    DurationContent, IndexContent, TextContent, TitleContent, TrackContent
} from "@components/music-table/music-table";

import { SupportedSources } from "./providerAuth";

export type Track = {
  id: string;
  isSelected: WritableSignal<boolean>;
  type: string;
  platform: SupportedSources;
  title: string;
  artist: {
    id: string;
    name: string;
    link: string;
  };
  album: {
    id: string;
    title: string;
    link: string;
  };
  duration_ms: number;
  link: string;
  preview: string | undefined;
  picture: string;
  release_date: string;
};

export type TrackResponse = Omit<Track, "isSelected">;

export type TracksResponse = TrackResponse[];

export type TracksTableData = {
  id: string;
  isSelected: WritableSignal<boolean>;
  content: [
    IndexContent,
    TitleContent,
    TextContent,
    TextContent,
    TrackContent,
    DurationContent,
  ];
};

File: backend/internal/provider/spotify/track.go
package spotify

type Track struct {
	ID           string       `json:"id"`
	Type         string       `json:"type"`
	Name         string       `json:"name"`
	Artists      []*Artist    `json:"artists"`
	Album        *Album       `json:"album"`
	DurationMS   int          `json:"duration_ms"`
	ExternalURLs ExternalURLs `json:"external_urls"`
	PreviewURL   string       `json:"preview_url"`
}

type MeTracksResponse struct {
	Items []struct {
		Track Track `json:"track"`
	}
}
